kythe/go/serving/pipeline/filetree.go

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pipeline
+
+import (
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"sort"
+
+	"kythe.io/kythe/go/serving/pipeline/nodes"
+	"kythe.io/kythe/go/util/kytheuri"
+
+	"github.com/apache/beam/sdks/go/pkg/beam"
+
+	srvpb "kythe.io/kythe/proto/serving_go_proto"
+	spb "kythe.io/kythe/proto/storage_go_proto"
+)
+
+func init() {
+	beam.RegisterFunction(addCorpusRootsKey)
+	beam.RegisterFunction(fileToCorpusRoot)
+	beam.RegisterFunction(fileToDirectories)
+
+	beam.RegisterType(reflect.TypeOf((*combineCorpusRoots)(nil)).Elem())
+	beam.RegisterType(reflect.TypeOf((*combineDirectories)(nil)).Elem())
+}
+
+func (k *KytheBeam) getFileVNames() beam.PCollection {
+	if k.fileVNames.IsValid() {
+		return k.fileVNames
+	}
+	k.fileVNames = beam.DropValue(k.s, beam.Seq(k.s, k.nodes, &nodes.Filter{
+		FilterByKind: []string{"file"},
+		IncludeFacts: []string{},
+		IncludeEdges: []string{},
+	}, moveSourceToKey))
+	return k.fileVNames
+}
+
+// CorpusRoots returns the single *srvpb.CorpusRoots key-value for the Kythe
+// FileTree service.  The beam.PCollection has elements of type KV<string,
+// *srvpb.CorpusRoots>.
+func (k *KytheBeam) CorpusRoots() beam.PCollection {
+	s := k.s.Scope("CorpusRoots")
+	files := k.getFileVNames()
+	return beam.ParDo(s, addCorpusRootsKey,
+		beam.Combine(s, &combineCorpusRoots{}, beam.ParDo(s, fileToCorpusRoot, files)))
+}
+
+// Directories returns a Kythe *srvpb.FileDirectory table for the Kythe FileTree
+// service.  The beam.PCollection has elements of type KV<string,
+// *srvpb.FileDirectory>.
+func (k *KytheBeam) Directories() beam.PCollection {
+	s := k.s.Scope("Directories")
+	files := k.getFileVNames()
+	return beam.CombinePerKey(s, &combineDirectories{}, beam.ParDo(s, fileToDirectories, files))
+}
+
+func addCorpusRootsKey(val beam.T) (string, beam.T) { return "dirs:corpusRoots", val }
+
+func fileToDirectories(file *spb.VName, emit func(string, *srvpb.FileDirectory)) {
+	path := filepath.Clean(filepath.Join("/", file.GetPath()))[1:]
+	dir := &spb.VName{
+		Corpus: file.Corpus,
+		Root:   file.Root,
+		Path:   filepath.Dir(path),
+	}
+	if dir.Path == "." {
```

Comment:
Do we have any interest in supporting ".." (or for that matter "...") in this context?

Whether we do or not, consider making this conversion a function so that we can use it again below (line 93).

Reply:
The above `Clean` handles those.

Reply:
Clean handles ".." and ".", but not "...".  Not sure whether we would ever encounter the latter in this context, though, so it may be a non-issue.

I'd still recommend using a function for this conversion, so that the code will be more readable/clean:

    func currentAsEmpty(string path) string {
      if path == "." {
        return ""
      }
      return path
    }
    ...
    dir.Path = currentAsEmpty(filePath.Dir(dir.Path))
    // etc.

but I don't insist on it.

Reply:
✓